### PR TITLE
Detect whether action view branch was deleted

### LINKED ISSUE #32761: Action view branch button doesn't check for deleted branch
### Description

image
Clicking on this button when branch was deleted show. In other pages branch existance is checked

### Gitea Version

trunk

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

docker

### Database

None

### PR DESCRIPTION
**Pull request by lunny**
Fix #32761 

图片

### CODE REVIEW

File: routers/web/repo/actions/view.go [hunk 1]
```diff
@@ -236,6 +237,16 @@ func ViewPost(ctx *context_module.Context) {
 		Name: run.PrettyRef(),
 		Link: run.RefLink(),
 	}
+	refName := git.RefName(run.Ref)
+	if refName.IsBranch() {
+		b, err := git_model.GetBranch(ctx, ctx.Repo.Repository.ID, refName.ShortName())
+		if err != nil && !git_model.IsErrBranchNotExist(err) {
+			log.Error("GetBranch: %v", err)
+		} else if git_model.IsErrBranchNotExist(err) || (b != nil && b.IsDeleted) {
+			branch.Link = "" // remove the link of branch if it doesn't exist'
```
Comment (metiftikci) [inline routers/web/repo/actions/view.go]: is not better to give link to ref?

**Comment by yp05327** [comment]
Not do it in `RefLink()`?

File: routers/web/repo/actions/view.go (hunk 1 above)
Comment (lunny) [reply routers/web/repo/actions/view.go]: 2b6a605aec4c83c1ff1e9dfb056edff54fd17381

**Comment by lunny** [comment]
> Not do it in `RefLink()`?
> 
> In `runs_list.tmpl`, it is also used. So I guess in runs list page, you will get similar issue. image

Done in 890863fe36138c17784530e3a7315944a95651e2

**Review by Zettat123 — APPROVED** [review]

File: routers/web/repo/actions/actions.go [hunk 2]
```diff
@@ -245,6 +245,11 @@ func List(ctx *context.Context) {
 		return
 	}
 
+	if err := loadIsRefDeleted(ctx, runs); err != nil {
+		ctx.ServerError("LoadIsRefDeleted", err)
```
Comment (yp05327) [inline routers/web/repo/actions/actions.go]: Maybe there's no need to return server error here.

Comment (lunny) [reply routers/web/repo/actions/actions.go]: 3a0676099d

File: models/git/branch.go [hunk 3]
```diff
@@ -174,6 +174,13 @@ func GetBranches(ctx context.Context, repoID int64, branchNames []string) ([]*Br
 	return branches, db.GetEngine(ctx).Where("repo_id=?", repoID).In("name", branchNames).Find(&branches)
 }
 
+func GetExistBranches(ctx context.Context, repoID int64, branchNames []string) ([]*Branch, error) {
```
Comment (delvh) [inline models/git/branch.go]: `GetExistingBranches`?

**Review by delvh — APPROVED** [review]

File: web_src/js/components/RepoActionView.vue [hunk 4]
```diff
@@ -429,7 +429,12 @@ export function initRepositoryActionView() {
           <a class="muted" :href="run.commit.pusher.link">{{ run.commit.pusher.displayName }}</a>
         </template>
         <span class="ui label tw-max-w-full" v-if="run.commit.shortSHA">
-          <a class="gt-ellipsis" :href="run.commit.branch.link">{{ run.commit.branch.name }}</a>
+          <template v-if="run.commit.branch.isDeleted">
+            <span class="gt-ellipsis tw-line-through" :data-tooltip-content="run.commit.branch.name">{{ run.commit.branch.name }}</span>
+          </template>
+          <template v-else>
+            <a class="gt-ellipsis" :href="run.commit.branch.link" :data-tooltip-content="run.commit.branch.name">{{ run.commit.branch.name }}</a>
+          </template>
```
Comment (wxiaoguang) [inline web_src/js/components/RepoActionView.vue]:
```suggestion
          <span v-if="run.commit.branch.isDeleted" class="gt-ellipsis tw-line-through" :data-tooltip-content="run.commit.branch.name">{{ run.commit.branch.name }}</span>
          <a v-else class="gt-ellipsis" :href="run.commit.branch.link" :data-tooltip-content="run.commit.branch.name">{{ run.commit.branch.name }}</a>
```

File: templates/repo/actions/runs_list.tmpl [hunk 5]
```diff
@@ -28,9 +28,13 @@
...
 				{{if .RefLink}}
-					<a class="ui label run-list-ref gt-ellipsis" href="{{.RefLink}}">{{.PrettyRef}}</a>
+					{{if .IsRefDeleted}}
+						<span class="ui label run-list-ref gt-ellipsis tw-line-through" data-tooltip-content="{{.PrettyRef}}">{{.PrettyRef}}</span>
+					{{else}}
+						<a class="ui label run-list-ref gt-ellipsis" href="{{.RefLink}}" data-tooltip-content="{{.PrettyRef}}">{{.PrettyRef}}</a>
+					{{end}}
 				{{else}}
-					<span class="ui label run-list-ref gt-ellipsis">{{.PrettyRef}}</span>
+					<span class="ui label run-list-ref gt-ellipsis" data-tooltip-content="{{.PrettyRef}}">{{.PrettyRef}}</span>
```
Comment (wxiaoguang) [inline templates/repo/actions/runs_list.tmpl]: It seems that `RefLink` could never be empty, so the `{{if .RefLink}}` could be removed? Then it also matches the Vue logic: only check `IsDeleted`

File: routers/web/repo/actions/actions.go [hunk 6]
```diff
@@ -267,6 +271,36 @@ func List(ctx *context.Context) {
...
+// TODO: move this function to models/actions/run_list.go but now it will result in a circular import.
+func loadIsRefDeleted(ctx *context.Context, runs actions_model.RunList) error {
+	branchRuns := make(map[string][]*actions_model.ActionRun)
+	branches := make(container.Set[string], len(runs))
+	for _, run := range runs {
+		refName := git.RefName(run.Ref)
+		if refName.IsBranch() {
+			branchName := refName.ShortName()
+			run.IsRefDeleted = true // assume it's deleted then if it's found in the database it's not deleted
+			branchRuns[branchName] = append(branchRuns[branchName], run)
```
Comment (wxiaoguang) [inline routers/web/repo/actions/actions.go]: Is it right? What if a branch starts 2 runs?

Comment (wxiaoguang) [reply routers/web/repo/actions/actions.go]: And since the logic seems fragile, I think it needs tests

File: models/git/branch.go (hunk 3 above)
Comment (wxiaoguang) [reply models/git/branch.go]: Agree, it needs a better name

File: routers/web/repo/actions/actions.go [hunk 7]
```diff
@@ -267,6 +271,36 @@ func List(ctx *context.Context) {
...
+	branchInfos, err := git_model.GetExistBranches(ctx, ctx.Repo.Repository.ID, branches.Values())
+	if err != nil {
+		return err
+	}
+	for _, branchInfo := range branchInfos {
+		branchRun := branchRuns[branchInfo.Name]
+		for _, br := range branchRun {
+			br.IsRefDeleted = false
+		}
+	}
```
Comment (wxiaoguang) [inline routers/web/repo/actions/actions.go:298]: Actually I think the logic should be like this:

```
for runs {
    runBranches.Add(run.branch)
}
existingBranches := GetExistingBranchNamesSet(runBranches)
for runs {
    run.IsDeleted = !existingBranches.Containts(runs.branch)
}

File: routers/web/repo/actions/actions.go (hunk 6 above)
Comment (lunny) [reply routers/web/repo/actions/actions.go]: > Is it right? What if a branch starts 2 runs?

A branch could start multiple runs, feature branches maybe updated multiple times and main branches may merge many commits. All these behaviours will trigger action runs.

File: routers/web/repo/actions/actions.go (hunk 7 above)
Comment (lunny) [reply routers/web/repo/actions/actions.go:298]: de6160a

File: routers/web/repo/actions/actions.go (hunk 6 above)
Comment (lunny) [reply routers/web/repo/actions/actions.go]: de6160a

File: web_src/js/components/RepoActionView.vue (hunk 4 above)
Comment (lunny) [reply web_src/js/components/RepoActionView.vue]: de6160a

File: models/git/branch.go (hunk 3 above)
Comment (lunny) [reply models/git/branch.go]: de6160a

**Pull request merged**